core/controllers/domain_objects_validator.py

```diff
@@ -178,3 +178,13 @@ def validate_aggregated_stats(aggregated_stats):
                 raise base.BaseHandler.InvalidInputException(
                     '%s not in state stats mapping of %s in aggregated '
                     'stats dict.' % (state_stats_property, state_name))
+
+
+def validate_params_dict(params):
+    """validates params data type
+
+    Args:
+        params: dict. Data that needs to be validated.
+    """
+    if not isinstance(params, dict):
+        raise Exception('Excepted dict, received %s' % params)
```

Comment:
I was asking to add a validation inorder to expect the type of the params to be dict, as you have done here.
Also verify that the keys in params dict are of string type.

Also can you modify the type annotation [here](https://github.com/oppia/oppia/blob/develop/core/storage/statistics/gae_models.py#L453)

Reply:
Yes, I found in many places that it is taking string-type keys.
previously I tried to add rules for both keys and values individually and that was causing lots of e2e failures and it seemed like they were transformed into string types before passing into the validator.
Also, I did not find anything that violates [them](https://github.com/oppia/oppia/blob/4e3d204fad3be36873fea307ae9c56ae1cf51a26/core/storage/statistics/gae_models.py#L453). then what did you ask me to modify? Can you please explain

Reply:
I think you can ignore this and leave it for now

Reply:
Okay.